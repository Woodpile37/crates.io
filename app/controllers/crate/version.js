import Controller from '@ember/controller';
import { action } from '@ember/object';
import { inject as service } from '@ember/service';
import { tracked } from '@glimmer/tracking';

import { task } from 'ember-concurrency';
import { alias } from 'macro-decorators';

export default class CrateVersionController extends Controller {
  @service session;

  get downloadsContext() {
    return this.requestedVersion ? this.currentVersion : this.crate;
  }

  @tracked stackedGraph = true;

  @action setStackedGraph() {
    this.stackedGraph = true;
  }

  @action setUnstackedGraph() {
    this.stackedGraph = false;
  }

  @alias('downloadsContext.version_downloads.content') downloads;
  @alias('model.crate') crate;
  @alias('model.requestedVersion') requestedVersion;
  @alias('model.version') currentVersion;

  get isOwner() {
    return this.crate.owner_user.findBy('id', this.session.currentUser?.id);
  }

  @alias('loadReadmeTask.last.value') readme;

  loadReadmeTask = task(async () => {
    let version = this.currentVersion;

    let readme = version.loadReadmeTask.lastSuccessful
      ? version.loadReadmeTask.lastSuccessful.value
      : await version.loadReadmeTask.perform();

    if (typeof document !== 'undefined') {
      setTimeout(() => {
        let e = document.createEvent('CustomEvent');
        e.initCustomEvent('hashchange', true, true);
        window.dispatchEvent(e);
      });
    }

    return readme;
  });
}
