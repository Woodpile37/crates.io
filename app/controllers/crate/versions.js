import Controller from '@ember/controller';
import { tracked } from '@glimmer/tracking';

export default class SearchController extends Controller {
  queryParams = ['sort'];

  @tracked sort;

  get currentSortBy() {
    return this.sort === 'semver' ? 'SemVer' : 'Date';
  }

  get sortedVersions() {
    let versions = this.model.versions.toArray();

    return this.sort === 'semver'
      ? versions.sort(compareBySemver)
      : versions.sort((a, b) => b.created_at - a.created_at);
  }
}

function compareBySemver(a, b) {
  let aSemver = a.semver;
  let bSemver = b.semver;

  if (aSemver === bSemver) {
    return b.created_at - a.created_at;
  } else if (aSemver === null) {
    return 1;
  } else if (bSemver === null) {
    return -1;
  } else {
    return bSemver.compare(aSemver);
  }
}
