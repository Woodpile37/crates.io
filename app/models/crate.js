import Model, { attr, hasMany } from '@ember-data/model';

import { customAction } from '../utils/custom-action';

export default class Crate extends Model {
  @attr name;
  @attr downloads;
  @attr recent_downloads;
  @attr('date') created_at;
  @attr('date') updated_at;
  @attr max_version;
  @attr max_stable_version;
  @attr newest_version;

  @attr description;
  @attr homepage;
  @attr documentation;
  @attr repository;

  @hasMany('versions', { async: true }) versions;

  @hasMany('teams', { async: true }) owner_team;
  @hasMany('users', { async: true }) owner_user;
  @hasMany('version-download', { async: true }) version_downloads;
  @hasMany('keywords', { async: true }) keywords;
  @hasMany('categories', { async: true }) categories;
  @hasMany('dependency', { async: true }) reverse_dependencies;

  /**
   * This is the default version that will be shown when visiting the crate
   * details page. Note that this can be `undefined` if all versions of the crate
   * have been yanked.
   * @return {string}
   */
  get defaultVersion() {
    if (this.max_stable_version) {
      return this.max_stable_version;
    }
    if (this.max_version && this.max_version !== '0.0.0') {
      return this.max_version;
    }
  }

  get owners() {
    let teams = this.owner_team.toArray() ?? [];
    let users = this.owner_user.toArray() ?? [];
    return [...teams, ...users];
  }

  async follow() {
    return await customAction(this, { method: 'PUT', path: 'follow' });
  }

  async unfollow() {
    return await customAction(this, { method: 'DELETE', path: 'follow' });
  }

  async inviteOwner(username) {
    let response = await customAction(this, { method: 'PUT', path: 'owners', data: { owners: [username] } });
    if (response.ok) {
      return response;
    } else {
      throw response;
    }
  }

  async removeOwner(username) {
    let response = await customAction(this, { method: 'DELETE', path: 'owners', data: { owners: [username] } });
    if (response.ok) {
      return response;
    } else {
      throw response;
    }
  }
}
