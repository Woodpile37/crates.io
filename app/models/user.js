import Model, { attr } from '@ember-data/model';
import { inject as service } from '@ember/service';

import { customAction } from '../utils/custom-action';

export default class User extends Model {
  @service store;

  @attr email;
  @attr email_verified;
  @attr email_verification_sent;
  @attr name;
  @attr login;
  @attr avatar;
  @attr url;
  @attr kind;
  @attr admin;

  async stats() {
    return await customAction(this, { method: 'GET', path: 'stats' });
  }

  async changeEmail(email) {
    await customAction(this, { method: 'PUT', data: { user: { email } } });

    this.store.pushPayload({
      user: {
        id: this.id,
        email,
        email_verified: false,
        email_verification_sent: true,
      },
    });
  }

  async resendVerificationEmail() {
    return await customAction(this, { method: 'PUT', path: 'resend' });
  }
}
