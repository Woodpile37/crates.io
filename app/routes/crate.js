import { NotFoundError } from '@ember-data/adapter/error';
import Route from '@ember/routing/route';
import { inject as service } from '@ember/service';

export default class CrateRoute extends Route {
  @service headData;
  @service router;
  @service store;

  async model(params, transition) {
    let crateName = params.crate_id;

    try {
      return await this.store.findRecord('crate', crateName);
    } catch (error) {
      if (error instanceof NotFoundError) {
        let title = `${crateName}: Crate not found`;
        this.router.replaceWith('catch-all', { transition, error, title });
      } else {
        let title = `${crateName}: Failed to load crate data`;
        this.router.replaceWith('catch-all', { transition, error, title, tryAgain: true });
      }
    }
  }

  setupController(controller, model) {
    super.setupController(...arguments);
    this.headData.crate = model;
  }

  resetController() {
    super.resetController(...arguments);
    this.headData.crate = null;
  }
}
