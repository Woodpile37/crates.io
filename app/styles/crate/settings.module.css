.email-form {
    display: flex;
    justify-content: space-between;
    align-items: center;
    flex-wrap: wrap;
    gap: var(--space-s);
    padding: var(--space-s) var(--space-m);
    background-color: white;
    border-radius: var(--space-3xs);
    box-shadow: 0 1px 3px hsla(51, 90%, 42%, .35);
}

.email-input-label {
    font-weight: bold;
}

.email-input {
    width: 400px;
}

.submit-button {
    composes: small yellow-button from '../shared/buttons.module.css';
}

.list {
    background-color: white;
    border-radius: var(--space-3xs);
    box-shadow: 0 1px 3px hsla(51, 90%, 42%, .35);

    > * {
        padding: var(--space-s) var(--space-m);
        display: flex;
        justify-content: space-between;
        align-items: center;
        flex-wrap: wrap;
    }

    > * + * {
        border-top: 1px solid hsla(51, 90%, 42%, .25);
    }
}

.email-column {
    width: 25%;
    color: var(--main-color-light);
}

.remove-button {
    composes: small yellow-button from '../shared/buttons.module.css';
}
