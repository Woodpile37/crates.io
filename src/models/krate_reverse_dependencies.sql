SELECT
    dependencies.*, crate_downloads, crate_name, total
FROM (
    -- Apply pagination to the crates
    SELECT *, COUNT(*) OVER () as total FROM (
        SELECT
            crates.downloads AS crate_downloads,
            crates.name AS crate_name,
            versions.id AS version_id
        FROM
        -- We only want the crates whose *max* version is dependent, so we join on a
        -- subselect that includes the versions with their ordinal position
        (
            SELECT DISTINCT ON (crate_id)
               crate_id, semver_no_prerelease, id
            FROM versions
            WHERE NOT yanked
            ORDER BY
                crate_id,
                semver_no_prerelease DESC NULLS LAST,
                id DESC
        ) versions
        INNER JOIN crates
          ON crates.id = versions.crate_id
        WHERE versions.id IN (SELECT version_id FROM dependencies WHERE crate_id = $1)
    ) c
    ORDER BY
        crate_downloads DESC,
        crate_name ASC
) crates
-- Multiple dependencies can exist, we only want first one
CROSS JOIN LATERAL (
    SELECT dependencies.*
    FROM dependencies
    WHERE dependencies.crate_id = $1 AND dependencies.version_id = crates.version_id
    ORDER BY id ASC
    LIMIT 1
) dependencies
OFFSET $2
LIMIT $3
