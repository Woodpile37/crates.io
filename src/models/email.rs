use chrono::NaiveDateTime;
use secrecy::SecretString;

use crate::models::User;
use crate::schema::emails;

#[derive(Debug, Queryable, Identifiable, Associations)]
#[diesel(belongs_to(User))]
pub struct Email {
    pub id: i32,
    pub user_id: i32,
    pub email: String,
    pub verified: bool,
    #[diesel(deserialize_as = String, serialize_as = String)]
    pub token: SecretString,
    pub token_generated_at: Option<NaiveDateTime>,
}

#[derive(Debug, Insertable, AsChangeset)]
#[diesel(table_name = emails, check_for_backend(diesel::pg::Pg))]
pub struct NewEmail<'a> {
    pub user_id: i32,
    pub email: &'a str,
}
