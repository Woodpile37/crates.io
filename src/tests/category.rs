use crate::util::insta::assert_yaml_snapshot;
use crate::{
    builders::CrateBuilder, new_category, util::MockAnonymousUser, RequestHelper, TestApp,
};
use cargo_registry::models::Category;
use serde_json::Value;

#[test]
fn index() {
    let (app, anon) = TestApp::init().empty();

    // List 0 categories if none exist
    let json: Value = anon.get("/api/v1/categories").good();
    assert_yaml_snapshot!(json);

    // Create a category and a subcategory
    app.db(|conn| {
        new_category("foo", "foo", "Foo crates")
            .create_or_update(conn)
            .unwrap();
        new_category("foo::bar", "foo::bar", "Bar crates")
            .create_or_update(conn)
            .unwrap();
    });

    // Only the top-level categories should be on the page
    let json: Value = anon.get("/api/v1/categories").good();
    assert_yaml_snapshot!(json, {
        ".categories[].created_at" => "[datetime]",
    });
}

#[test]
fn show() {
    let (app, anon) = TestApp::init().empty();
    let url = "/api/v1/categories/foo-bar";

    // Return not found if a category doesn't exist
    anon.get(url).assert_not_found();

    // Create a category and a subcategory
    app.db(|conn| {
        assert_ok!(new_category("Foo Bar", "foo-bar", "Foo Bar crates").create_or_update(conn));
        assert_ok!(
            new_category("Foo Bar::Baz", "foo-bar::baz", "Baz crates").create_or_update(conn)
        );
    });

    // The category and its subcategories should be in the json
    let json: Value = anon.get(url).good();
    assert_yaml_snapshot!(json, {
        ".**.created_at" => "[datetime]",
    });
}

#[test]
#[allow(clippy::cognitive_complexity)]
fn update_crate() {
    // Convenience function to get the number of crates in a category
    fn count(anon: &MockAnonymousUser, category: &str) -> usize {
        let json = anon.show_category(category);
        json.category.crates_cnt as usize
    }

    let (app, anon, user) = TestApp::init().with_user();
    let user = user.as_model();

    app.db(|conn| {
        assert_ok!(new_category("cat1", "cat1", "Category 1 crates").create_or_update(conn));
        assert_ok!(
            new_category("Category 2", "category-2", "Category 2 crates").create_or_update(conn)
        );
        let krate = CrateBuilder::new("foo_crate", user.id).expect_build(conn);

        // Updating with no categories has no effect
        Category::update_crate(conn, &krate, &[]).unwrap();
        assert_eq!(count(&anon, "cat1"), 0);
        assert_eq!(count(&anon, "category-2"), 0);

        // Happy path adding one category
        Category::update_crate(conn, &krate, &["cat1"]).unwrap();
        assert_eq!(count(&anon, "cat1"), 1);
        assert_eq!(count(&anon, "category-2"), 0);

        // Replacing one category with another
        Category::update_crate(conn, &krate, &["category-2"]).unwrap();
        assert_eq!(count(&anon, "cat1"), 0);
        assert_eq!(count(&anon, "category-2"), 1);

        // Removing one category
        Category::update_crate(conn, &krate, &[]).unwrap();
        assert_eq!(count(&anon, "cat1"), 0);
        assert_eq!(count(&anon, "category-2"), 0);

        // Adding 2 categories
        Category::update_crate(conn, &krate, &["cat1", "category-2"]).unwrap();
        assert_eq!(count(&anon, "cat1"), 1);
        assert_eq!(count(&anon, "category-2"), 1);

        // Removing all categories
        Category::update_crate(conn, &krate, &[]).unwrap();
        assert_eq!(count(&anon, "cat1"), 0);
        assert_eq!(count(&anon, "category-2"), 0);

        // Attempting to add one valid category and one invalid category
        let invalid_categories =
            Category::update_crate(conn, &krate, &["cat1", "catnope"]).unwrap();
        assert_eq!(invalid_categories, vec!["catnope"]);
        assert_eq!(count(&anon, "cat1"), 1);
        assert_eq!(count(&anon, "category-2"), 0);

        // Does not add the invalid category to the category list
        // (unlike the behavior of keywords)
        let json = anon.show_category_list();
        assert_eq!(json.categories.len(), 2);
        assert_eq!(json.meta.total, 2);

        // Attempting to add a category by display text; must use slug
        Category::update_crate(conn, &krate, &["Category 2"]).unwrap();
        assert_eq!(count(&anon, "cat1"), 0);
        assert_eq!(count(&anon, "category-2"), 0);

        // Add a category and its subcategory
        assert_ok!(new_category("cat1::bar", "cat1::bar", "bar crates").create_or_update(conn));
        Category::update_crate(conn, &krate, &["cat1", "cat1::bar"]).unwrap();

        assert_eq!(count(&anon, "cat1"), 1);
        assert_eq!(count(&anon, "cat1::bar"), 1);
        assert_eq!(count(&anon, "category-2"), 0);
    });
}

#[test]
fn category_slugs_returns_all_slugs_in_alphabetical_order() {
    let (app, anon) = TestApp::init().empty();
    app.db(|conn| {
        new_category("Foo", "foo", "For crates that foo")
            .create_or_update(conn)
            .unwrap();
        new_category("Bar", "bar", "For crates that bar")
            .create_or_update(conn)
            .unwrap();
    });

    let response: Value = anon.get("/api/v1/category_slugs").good();
    assert_yaml_snapshot!(response);
}
