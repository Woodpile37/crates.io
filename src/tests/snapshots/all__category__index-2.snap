---
source: src/tests/category.rs
assertion_line: 33
expression: json
---
categories:
  - category: foo
    crates_cnt: 0
    created_at: "[datetime]"
    description: Foo crates
    id: foo
    slug: foo
meta:
  total: 1

